Datastore file (csv) upload sometimes does not work for the first time

Sometimes after a fresh install of the Monitor to the Tomcat the datastore file upload (tested with csv) does not work. The second attempt (immediately after the error) always works fine. 

1. Install new Monitor.war to the tomcat. 
2. Open "Datastores" page. 
3. Click "Register datastore". 
4. Select CSV & Next. 
5. "I want to upload the file" & Browse file & select & upload & Next. 
6. "Copy it to a location in the repository" & Next. 
7. Server reported HTTP error: 500. 

![select-csv](https://cloud.githubusercontent.com/assets/13213915/21006304/a1731016-bd38-11e6-921a-2a11d06d0488.png)
![upload-is-fine](https://cloud.githubusercontent.com/assets/13213915/21006306/a17864ee-bd38-11e6-9ec8-6ae386e717f1.png)
![to-datastores](https://cloud.githubusercontent.com/assets/13213915/21006305/a17678dc-bd38-11e6-9758-bcd1d8dcf0bf.png)
![error](https://cloud.githubusercontent.com/assets/13213915/21006307/a18017a2-bd38-11e6-8d0f-49a6e5ab8e9e.png)

```
11:22:00.861 [http-nio-8080-exec-2] WARN  o.d.monitor.server.SecureGwtServlet - Unexpected exception occurred in GWT servlet: com.google.gwt.user.server.rpc.UnexpectedException
com.google.gwt.user.server.rpc.UnexpectedException: Service method 'public abstract org.datacleaner.monitor.shared.model.WizardPage org.datacleaner.monitor.shared.WizardService.nextPage(org.datacleaner.monitor.shared.model.TenantIdentifier,org.datacleaner.monitor.shared.model.WizardSessionIdentifier,java.util.Map) throws org.datacleaner.monitor.shared.model.DCUserInputException' threw an unexpected exception: java.lang.IllegalStateException: java.io.FileNotFoundException: /home/kuba/apps/apache-tomcat-8.0.30/temp/upload__55e06d1d_158ddec3d6d__8000_00000003.tmp (No such file or directory)
	at com.google.gwt.user.server.rpc.RPC.encodeResponseForFailure(RPC.java:415) ~[gwt-servlet-2.7.0.jar:na]
	at com.google.gwt.user.server.rpc.RPC.invokeAndEncodeResponse(RPC.java:605) ~[gwt-servlet-2.7.0.jar:na]
	at com.google.gwt.user.server.rpc.RemoteServiceServlet.processCall(RemoteServiceServlet.java:333) ~[gwt-servlet-2.7.0.jar:na]
	at com.google.gwt.user.server.rpc.RemoteServiceServlet.processCall(RemoteServiceServlet.java:303) ~[gwt-servlet-2.7.0.jar:na]
	at com.google.gwt.user.server.rpc.RemoteServiceServlet.processPost(RemoteServiceServlet.java:373) ~[gwt-servlet-2.7.0.jar:na]
	at com.google.gwt.user.server.rpc.AbstractRemoteServiceServlet.doPost(AbstractRemoteServiceServlet.java:62) ~[gwt-servlet-2.7.0.jar:na]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:648) [servlet-api.jar:na]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729) [servlet-api.jar:na]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291) [catalina.jar:8.0.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.30]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) [tomcat-websocket.jar:8.0.30]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [catalina.jar:8.0.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.30]
	at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:205) [rewrite-servlet-2.0.12.Final.jar:2.0.12.Final]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [catalina.jar:8.0.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.30]
	at org.datacleaner.monitor.server.GwtCacheControlFilter.doFilter(GwtCacheControlFilter.java:70) [DataCleaner-monitor-services-5.1.6-SNAPSHOT.jar:na]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [catalina.jar:8.0.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.30]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:103) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:113) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:146) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:154) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:45) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:150) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:199) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:110) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:50) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160) [spring-security-web-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:344) [spring-web-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:261) [spring-web-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239) [catalina.jar:8.0.30]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206) [catalina.jar:8.0.30]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212) [catalina.jar:8.0.30]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106) [catalina.jar:8.0.30]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502) [catalina.jar:8.0.30]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141) [catalina.jar:8.0.30]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79) [catalina.jar:8.0.30]
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616) [catalina.jar:8.0.30]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88) [catalina.jar:8.0.30]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:521) [catalina.jar:8.0.30]
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1096) [tomcat-coyote.jar:8.0.30]
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674) [tomcat-coyote.jar:8.0.30]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500) [tomcat-coyote.jar:8.0.30]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456) [tomcat-coyote.jar:8.0.30]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_72]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_72]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:8.0.30]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_72]
Caused by: java.lang.IllegalStateException: java.io.FileNotFoundException: /home/kuba/apps/apache-tomcat-8.0.30/temp/upload__55e06d1d_158ddec3d6d__8000_00000003.tmp (No such file or directory)
	at org.apache.metamodel.util.FileHelper.getInputStream(FileHelper.java:444) ~[MetaModel-core-4.5.4.jar:4.5.4]
	at org.apache.metamodel.util.FileHelper.copy(FileHelper.java:415) ~[MetaModel-core-4.5.4.jar:4.5.4]
	at org.datacleaner.monitor.server.wizard.CsvDatastoreWizardSession$1$1.nextPageController(CsvDatastoreWizardSession.java:82) ~[DataCleaner-monitor-csv-datastore-wizard-5.1.6-SNAPSHOT.jar:na]
	at org.datacleaner.monitor.server.wizard.CsvDatastoreLocationWizardPage.nextPageController(CsvDatastoreLocationWizardPage.java:140) ~[DataCleaner-monitor-csv-datastore-wizard-5.1.6-SNAPSHOT.jar:na]
	at org.datacleaner.monitor.server.dao.WizardDaoImpl.nextPage(WizardDaoImpl.java:101) ~[DataCleaner-monitor-services-5.1.6-SNAPSHOT.jar:na]
	at org.datacleaner.monitor.server.WizardServiceImpl.nextPage(WizardServiceImpl.java:197) ~[DataCleaner-monitor-services-5.1.6-SNAPSHOT.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_72]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_72]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_72]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_72]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317) ~[spring-aop-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190) ~[spring-aop-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) ~[spring-aop-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:64) ~[spring-security-core-3.2.9.RELEASE.jar:3.2.9.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207) ~[spring-aop-4.1.9.RELEASE.jar:4.1.9.RELEASE]
	at com.sun.proxy.$Proxy51.nextPage(Unknown Source) ~[na:na]
	at org.datacleaner.monitor.server.WizardServiceServlet.nextPage(WizardServiceServlet.java:86) ~[DataCleaner-monitor-services-5.1.6-SNAPSHOT.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_72]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_72]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_72]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_72]
	at com.google.gwt.user.server.rpc.RPC.invokeAndEncodeResponse(RPC.java:587) ~[gwt-servlet-2.7.0.jar:na]
	... 66 common frames omitted
Caused by: java.io.FileNotFoundException: /home/kuba/apps/apache-tomcat-8.0.30/temp/upload__55e06d1d_158ddec3d6d__8000_00000003.tmp (No such file or directory)
	at java.io.FileInputStream.open0(Native Method) ~[na:1.8.0_72]
	at java.io.FileInputStream.open(FileInputStream.java:195) ~[na:1.8.0_72]
	at java.io.FileInputStream.<init>(FileInputStream.java:138) ~[na:1.8.0_72]
	at org.apache.metamodel.util.FileHelper.getInputStream(FileHelper.java:442) ~[MetaModel-core-4.5.4.jar:4.5.4]
	... 88 common frames omitted
```

--- a/monitor/extensions/csv-datastore-wizard/src/main/resources/org/datacleaner/monitor/server/wizard/CsvDatastoreUploadOrExistingFileWizardPage.html
+++ b/monitor/extensions/csv-datastore-wizard/src/main/resources/org/datacleaner/monitor/server/wizard/CsvDatastoreUploadOrExistingFileWizardPage.html
@@ -23,9 +23,10 @@
 			want to upload the file to the server (max. 50mb)</label><br />
 
 		<div id="csv_file_upload_div">
-			<input type="file" name="csv_file_upload" /> <input type="button"
+			<input type="file" name="csv_file_upload"  onchange="window.enableUploadButton('uploadButton');"/> 
+			<input type="button" id="uploadButton"
 				value="upload now"
-				onclick="window.uploadFile('csv_file_upload_div');" />
+				onclick="window.uploadFile('csv_file_upload_div');" disabled/>
 		</div>
 	</div>
 </div>

--- a/monitor/extensions/excel-datastore-wizard/src/main/resources/org/datacleaner/monitor/server/wizard/ExcelDatastoreUploadOrExistingFileWizardPage.html
+++ b/monitor/extensions/excel-datastore-wizard/src/main/resources/org/datacleaner/monitor/server/wizard/ExcelDatastoreUploadOrExistingFileWizardPage.html
@@ -22,9 +22,11 @@
 			want to upload the file to the server</label><br />
 
 		<div id="excel_spreadsheet_upload_div">
-			<input type="file" name="excel_spreadsheet_upload" /> <input type="button"
-				value="upload now"
-				onclick="window.uploadFile('excel_spreadsheet_upload_div');" />
+			<input type="file" name="excel_spreadsheet_upload"
+				onchange="window.enableUploadButton('uploadButton');" /> <input
+				type="button" id="uploadButton" value="upload now"
+				onclick="window.uploadFile('excel_spreadsheet_upload_div');"
+				disabled />
 		</div>
 	</div>
 </div>

--- a/monitor/widgets/src/main/java/org/datacleaner/monitor/shared/widgets/FileUploadFunctionHandler.java
+++ b/monitor/widgets/src/main/java/org/datacleaner/monitor/shared/widgets/FileUploadFunctionHandler.java
@@ -41,6 +41,13 @@ import com.google.gwt.user.client.ui.RootPanel;
  * FileUploadServlet available in DC monitor.
  */
 public class FileUploadFunctionHandler {
+    
+    public static void enableUploadButton(final String buttonId) {
+        final Element element = Document.get().getElementById(buttonId);
+
+        GWT.log("Found file upload button element: " + element);
+        element.removeAttribute("disabled");
+    }
 
     public static void uploadFile(final String fileUploadElementId) {
         final Element element = Document.get().getElementById(fileUploadElementId);
@@ -134,4 +141,15 @@ public class FileUploadFunctionHandler {
     public static native void exportFileUploadFunction() /*-{
         $wnd.uploadFile = $entry(@org.datacleaner.monitor.shared.widgets.FileUploadFunctionHandler::uploadFile(Ljava/lang/String;));
     }-*/;
+    
+    /**
+     * Exports the "enableUploadButton(final String buttonId)" method as a function in the native
+     * javascript scope.
+     */
+    @SuppressWarnings("checkstyle:LineLength")
+    public static native void exportEnableUploadButtonFunction() /*-{
+        $wnd.enableUploadButton = $entry(@org.datacleaner.monitor.shared.widgets.FileUploadFunctionHandler::enableUploadButton(Ljava/lang/String;));
+    }-*/;
+    
+    
 }

--- a/monitor/widgets/src/main/java/org/datacleaner/monitor/wizard/AbstractWizardController.java
+++ b/monitor/widgets/src/main/java/org/datacleaner/monitor/wizard/AbstractWizardController.java
@@ -78,6 +78,7 @@ public abstract class AbstractWizardController<S extends WizardNavigationService
         _tenant = tenant;
 
         FileUploadFunctionHandler.exportFileUploadFunction();
+        FileUploadFunctionHandler.exportEnableUploadButtonFunction();
 
         _loadingIndicator = new LoadingIndicator();
         _wizardPanel.setContent(_loadingIndicator);

